gnd/src/main/java/com/google/android/gnd/repository/OfflineBaseMapRepository.java

```diff
@@ -186,4 +212,23 @@ public Completable addAreaAndEnqueue(LatLngBounds bounds) {
                     .filter(tileSource -> tileSource.getState() == TileSource.State.DOWNLOADED)
                     .collect(toImmutableSet()));
   }
+
+  /**
+   * Delete an offline base map and any tile sources associated with it that do not overlap with
+   * other offline base maps .
+   */
+  public Completable deleteArea(String offlineAreaId) {
+    return localDataStore
+        .getOfflineAreaById(offlineAreaId)
+        .flatMapMaybe(this::getIntersectingDownloadedTileSourcesOnce)
+        .flatMapPublisher(Flowable::fromIterable)
```

Comment:
Can we avoid casting to Flowable here to simplify a bit?

Reply:
Unless I'm missing something we'll either need to cast the initial iterable into a flowable or flatten a list of resulting  Completables using `merge`. I find converting an iterable to a stream of items and mapping over the stream to be more idiomatic and simpler to understand. For comparison, here's the current approach:

```java
.getOfflineAreaById(offlineAreaId)
        .flatMapMaybe(this::getIntersectingDownloadedTileSourcesOnce)
        .flatMapPublisher(Flowable::fromIterable) // convert List<TileSource> to a stream of <TileSource>
        .map(TileSource::decrementAreaCount) // from here on, just use straightforward RX stream operations to manipulate each item
        .flatMapCompletable(
            tile ->
                localDataStore
                    .updateTileSourceAreaCountByUrl(tile.getAreaCount(), tile.getUrl())
                    .andThen(localDataStore.deleteTileByUrl(tile)))
        .andThen(localDataStore.deleteOfflineArea(offlineAreaId));
```

versus the alternative explicit iteration over the iterable:

```java
Completable.merge(
  localdatastore.getOfflineAreaById(offlineAreaId)
        .flatMapMaybe(this::getIntersectingDownloadedTileSourcesOnce)
        .map(tileSources -> 
                    stream(tileSources).map( // we could have just converted this into an RX stream but instead we use a Java 8 stream...within a stream
                       tile ->
                          localDataStore
                          .updateTileSourceAreaCountByUrl(tile.getAreaCount(), tile.getUrl())
                          .andThen(localDataStore.deleteTileByUrl(tile))).collect(toImmutableList))) // list of Completable
        .andThen(localDataStore.deleteOfflineArea(offlineAreaId));
```

or something similar. I think the first, simply casting an iterable into terms of RX java streams has a few advantages:

- eliminate stream util code (since rx java operates on streams and provides conversions from iterables into streams, isn't somewhat silly for us to stream(listofThings).map...etc. within a stream, we're effectively using two different but equivalent stream languages (java 8 and RX) at different levels)?
- eliminate recollection and flattenting
- operate on a stream of values rather than a pseudo stream of a single value which is itself a streamable collection.

I think the interior stream approach makes sense when we want to continue talking about a collection of values in the RX stream (for instance, if we were collecting these after transforming them into ints) but when the final value is something like a Completable I think it makes more sense just to up-level the items to the RX stream level.

That said I'm not sure what the relative (if any) performance trade offs are between each approach, so this is mostly a stylistic preference. I just think thinking in terms of streams of tiles at a single level (RX) is simpler than having to think in terms of streams at two levels (the RX stream *and* the stream of tiles within that stream) and have to flatten an iterable of stream/observable types (which arguably I think you should only be doing when you have some more meaningful combination of multiple observables, such as observables that emit different items that need to be combined)

Reply:
@gino-m

Reply:
Sorry, I should have been more clear - I think unpacking the collection makes sense, I was wondering if we can use `Observable::fromIterable` here instead of `Flowable::fromIterable`

Reply:
oh! Indeed we can! And it's better since we've no real need for backpressure here.

Reply:
364195f